MH-13174 Simplify class MediaPackageElementFlavor

staubesv opened the pull request:
- Make default constructor private to ensure that type and subtype are never null
- Don't handle the cases type == null and subtype == null to simplify the boolean expressions
- Remove description and related functionality since unused
- Remove the ElementTypeEquivalent logic as unused

staubesv committed:
MH-13174 Simplify class MediaPackageElementFlavor

- Make default constructor private to ensure that type and subtype are never null
- Don't handle the cases type == null and subtype == null to simplify the boolean expressions
- Remove description and related functionality since unused
- Remove the ElementTypeEquivalent logic as unused

lkiesow commented on modules/common/src/main/java/org/opencastproject/mediapackage/MediaPackageElementFlavor.java:
```diff
@@ -81,10 +77,9 @@
    * @param description
```
This parameter does not exist any longer and breaks `mvn site`

staubesv replied:
Yep, oversight...

lkiesow commented on modules/common/src/main/java/org/opencastproject/mediapackage/MediaPackageElementFlavor.java:
```diff
@@ -392,52 +254,27 @@ public MediaPackageElementFlavor unmarshal(String str) throws Exception {
     int result = 1;
-    result = prime * result + ((subtype == null) ? 0 : subtype.hashCode());
-    result = prime * result + ((type == null) ? 0 : type.hashCode());
+    result = prime * result + subtype.hashCode();
```
We don't even need to initialize this before:

```suggestion
    int result = prime + subtype.hashCode();
```

lkiesow replied:
Thinking about it again, the whole method is just:

```java
public int hashCode() {
￼     return 961 + subtype.hashCode() + type.hashCode();
}
```

staubesv replied:
But this doesn't explain anymore that a prime is used to build the hash code... the code should be easy to understand - not necessarily as short as possible.

lkiesow replied:
True but irrelevant because it does not really matter that it's a prime number, does it?
In both cases, it's basically only a randomly chosen magic number.

But if you want to keep it, maybe:

```java
final int prime = 31;
return prime * prime + subtype.hashCode() + type.hashCode();
```

Because the `result` variable does not make it easier to read.

lkiesow replied:
Shortly looked into this. With Java7+, this seems to be what we really want to use instead of writing our own code (sorry for all the updates):

```java
return Objects.hash(subtype, type);
```

staubesv replied:
I did some quick research but it does not seem 100% clear whether Objects.hash is a good choice.

Maybe it is, but I don't really have time to look into this. While this code could be further simplified, it is not as terrible as some other code here was...

lkiesow commented on modules/common/src/main/java/org/opencastproject/mediapackage/MediaPackageElementFlavor.java:
```diff
@@ -392,52 +254,27 @@ public MediaPackageElementFlavor unmarshal(String str) throws Exception {
-      return false;
-    return true;
+  public boolean equals(Object other) {
+    return (other != null) && (other instanceof MediaPackageElementFlavor)
```
`instanceof` should already take care of the null check:

```suggestion
    return other instanceof MediaPackageElementFlavor
```

staubesv replied:
Thanks, didn't know that.

lkiesow reviewed (changes requested):
Appreciate the pull request.
Everything except the `mvn site` build error caused is purely optional.

lkiesow commented:
Small addition:

https://github.com/opencast/opencast/blob/e6998f2153224084aef7282c6492f73d7be8559a/modules/common/src/main/java/org/opencastproject/mediapackage/MediaPackageElementFlavor.java#L250-L253

This could also be just

```java
return parseFlavor(str); 
```

staubesv committed:
MH-13174 Simplify class MediaPackageElementFlavor (Round 2)

- Fixed wrong comments
- Remove unused method hasWildcard
- Some further code simplifications

staubesv commented:
@lkiesow 

I've updated the PR and did some further simplifications. Note that there would be more desirable work to do, e.g. removing the weird functional-style parseFlavor method (how many functional libraries does Opencast have?) which is just used by two WOHs. Probably the format() in the log could also be removed..

... anyway, already spent too much time on this and while there is more room for simplification, this pull request does simplify the class quite a bit ;-)

lkiesow commented:
Thanks for the patch. Merged.
